src/Compilers/CSharp/Portable/Symbols/Compilation_WellKnownMembers.cs

```diff
@@ -491,6 +491,57 @@ internal SynthesizedAttributeData SynthesizeDynamicAttribute(TypeSymbol type, in
             }
         }
 
+        internal SynthesizedAttributeData SynthesizeTupleNamesAttribute(TypeSymbol type)
+        {
+            Debug.Assert((object)type != null);
+            Debug.Assert(type.ContainsTuple());
+
+            var stringType = GetSpecialType(SpecialType.System_String);
+            Debug.Assert((object)stringType != null);
+            var names = TupleNamesEncoder.Encode(type, stringType);
+            var stringArray = ArrayTypeSymbol.CreateSZArray(stringType.ContainingAssembly, stringType, customModifiers: ImmutableArray<CustomModifier>.Empty);
+            var args = ImmutableArray.Create(new TypedConstant(stringArray, names));
+            return TrySynthesizeAttribute(WellKnownMember.System_Runtime_CompilerServices_TupleElementNamesAttribute__ctorTransformNames, args);
+        }
+
+        private static class TupleNamesEncoder
+        {
+            public static ImmutableArray<TypedConstant> Encode(TypeSymbol type, TypeSymbol stringType)
+            {
+                var namesBuilder = ArrayBuilder<string>.GetInstance();
+                type.VisitType((t, builder, isNested) => AddName(t, builder, isNested), namesBuilder);
+                Debug.Assert(namesBuilder.Any());
+
+                var constantsBuilder = ArrayBuilder<TypedConstant>.GetInstance(namesBuilder.Count);
+                foreach (var name in namesBuilder)
+                {
+                    constantsBuilder.Add(new TypedConstant(stringType, TypedConstantKind.Primitive, name));
+                }
+
+                namesBuilder.Free();
+                return constantsBuilder.ToImmutableAndFree();
+            }
+
+            private static bool AddName(TypeSymbol type, ArrayBuilder<string> namesBuilder, bool isNestedType)
+            {
+                if (type.IsTupleType)
+                {
+                    if (type.TupleElementNames.IsDefaultOrEmpty)
```

Comment:
I think we can assume that tuples that will make it to be encoded in metadata signatures will either have no element names or all elements will have non-null names, thus this can be "optimized" by adding just a single null representing "no names" case.
If this is NYI, perhaps need to add a workitem for this.